hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -487,8 +487,17 @@ struct TypeVisitor : OptimizerVisitor, visitor::PreOrder<bool, TypeVisitor> {
             case Stage::PRUNE_DECLS:
                 if ( ! _used.at(*type_id) ) {
                     HILTI_DEBUG(logging::debug::GlobalOptimizer, util::fmt("removing unused type '%s'", *type_id));
+
                     removeNode(p);
 
+                    if ( auto module_ = p.findParent<Module>() )
+                        // If this type was declared under a top-level module also clear the module declaration
+                        // cache. The cache will get repopulated the next time the module's declarations are
+                        // requested.
+                        //
+                        // TODO(bbannier): there has to be a better way to mutate the module.
+                        const_cast<Module&>(module_->get()).clearCache();
```

Comment:
Another thought: you don't neccessarily need to restrict this to modules and could create a helper functions similar to `Unit::_resetNodes` that just generally clears scopes & caches for an AST; and then call that from different places.

But fine to leave as is as well, as the need for this will go away anyways with the new AST code.

Reply:
Unfortunately this only works if we have the position of the module which is not the case here (but works for the other `const_cast` below). Here we call `findParent` to get the `Module` which always returns an optional ref of `const T`. :/